### PR TITLE
UTIL: ucc cpu arch specific functions

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Adding CPU architecture specific functions

## Why ?
Required to implement TL CUDA and TL SHM

## How ?
Copy of needed functions from UCS

### CODE REVIEW

File: src/utils/arch/aarch64/cpu.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
```
Comment (vspetrov) [inline src/utils/arch/aarch64/cpu.h:2]: 2021

**Review by vspetrov — APPROVED** [review]

**Review by shimmybalsam — APPROVED** [review]

**Review by shimmybalsam — COMMENTED** [review]
Please add isync function as well

File: src/utils/arch/aarch64/cpu.h (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/utils/arch/aarch64/cpu.h]: fixed

**Comment by Sergei-Lebedev** [comment]
> Please add isync function as well

added

File: src/utils/ucc_compiler_def.h [hunk 2]
```diff
@@ -25,6 +25,11 @@
 #define ucc_likely        ucs_likely
 #define ucc_unlikely      ucs_unlikely
 
+/**
+ * Prevent compiler from reordering instructions
+ */
+#define ucc_compiler_fence()       asm volatile(""::: "memory")
```
Comment (vspetrov) [inline src/utils/ucc_compiler_def.h:31]: @Sergei-Lebedev i think i was not right saying that Shimmy needs just a compiler instruction barrier. We need arch specific CPU instruction barrier.
in hcoll we use libocoms which defines isync:
power: https://github.com/uccs/libocoms/blob/4aa64d0105317badb244e7433ea29aa03a014507/ocoms/sys/powerpc/atomic.h#L34

arm: https://github.com/uccs/libocoms/blob/4aa64d0105317badb244e7433ea29aa03a014507/ocoms/sys/arm64/atomic.h#L76

@bureddy we need an instruction sync primitive for shared memory collectives (remember ppc data race) - does ucx/ucs have that or we need to copy paste from libocoms?

**Comment by manjugv** [comment]
@Sergei-Lebedev I'm missing context here. Why do you need to make a copy of these functions?

**Comment by Sergei-Lebedev** [comment]
> @Sergei-Lebedev I'm missing context here. Why do you need to make a copy of these functions?

currently those functions are not part of ucs headers and it's unlikely they ever be public

**Comment by manjugv** [comment]
> > @Sergei-Lebedev I'm missing context here. Why do you need to make a copy of these functions?
> 
> currently those functions are not part of ucs headers and it's unlikely they ever be public

Do you know why is it not public?

**Comment by manjugv** [comment]
@Sergei-Lebedev Once @bureddy confirms it is ready to go.

File: src/utils/ucc_compiler_def.h (hunk 2 above)
Comment (bureddy) [reply src/utils/ucc_compiler_def.h:31]: ucx/ucs  ucc_memory_cpu_load_fence() should be enough for pcc shared memory.

Comment (bureddy) [reply src/utils/ucc_compiler_def.h:31]: here was the requirement  on PPC
```

producer 
//store data
for( .............)
 shared_array[i] = 0;                                                                                                                  
ppc_write_fence();   //an lwsync is sufficient for DDR memory writes
//Notify consumers
shared_flag = 1;

Consumer
while (!shared_flag);    //blocking wait on the shared flag (ideally should be non-blocking)
isync();                             //Don't let the speculation unit move instructions ahead of the while loop check
//access the shared_array
int x = shared_array[y];     //read the updated shared state

```

**Review by bureddy — APPROVED** [review]

**Pull request merged**